### DISCUSSION TITLE
Performance considerations for calling recursive functions

### DISCUSSION THREAD

**Discussion by CreepGin** [sebastienros/jint#1266] [General]
I'm noticing a big perf hit when using Jint to call certain recursive JS functions. Specifically, I'm calling Preact's diff() function for DOM re-render. Everything works fine, but it's just quite slow even for very simple DOM hierarchy. And it seem most of the perf hit comes from GC allocations. 

I've been trying to reduce the perf hit for a few weeks now without much success. So I'm just wondering if anyone can bounce some ideas my way. Thanks!

**Comment by lahma** [comment]
Can you create a sample program that demonstrates the problem so it would be a base for possible performance tweaking on Jint's side?

**Reply by CreepGin** [reply]
Thanks @lahma  My case is a little unique as I'm using Unity's UI Toolkit as the DOM for Preact. So my current minimal testable sample code would be a Unity project. Would that still be useful to you?

**Reply by lahma** [reply]
That might be a bit hard to profile. I think this issue is related and has recursive repro sample: https://github.com/sebastienros/jint/issues/1159 .

**Reply by CreepGin** [reply]
Yes looks to be related. Thx for that info! I'll do more testing, especially with your new commits regarding the stack frame size.

**Reply by lahma** [reply]
Yes, please do, MyGet feed will always have the latest bits. One optimization might be trying to find a new function type that can be used to do a "lightweight" invocation - less arguments handling etc when AST is known to be "care free". Would you happen to have JS example how the invocation code looks like - nesting and function parameters?

**Reply by CreepGin** [reply]
Thanks, that's very helpful! And yes, I think I was just able to reproduce the preact diff() slowdown using the simple example @KurtGokhan used. I'll hop into that issue thread to further the investigation.
